Container runtime

yifan-gu opened the pull request:
rebased. Also worked for docker by adding a 'jump wire'.

Commit:
kubelet: add container runtime interface.

Commit:
kubelet: add container runtime cache.

Commit:
kubelet: add rocket implementation for container runtime interface.

Commit:
kubelet: add tests for rocket runtime.

Commit:
Godeps: update godeps for rocket.

Commit:
kubelet: Add new sync path for rocket.

Commit:
kubelet: add manageRocketPodLoop for pod_workers.go.

Commit:
kubelet: update the interface of newPodWorkers() to support rocket.

Commit:
cmd and kubelet: add jump wire for running rocket.

Commit:
kubelet: update cleanupOrphanedVolumes() to take only pods as input.

yifan-gu commented:
Current state and issues here. Feed back welcome! Thank you! @philips @bcwaldon @jonboulle @sym3tri 
https://docs.google.com/a/coreos.com/document/d/1-gMvxSuol_gXiIFodfzAzOIlz5_MYLT1HZMElP6Yd6w/edit

sym3tri commented on pkg/kubelet/kubelet.go:
```diff
@@ -1943,3 +1947,222 @@ func (kl *Kubelet) GetRootInfo(req *cadvisorApi.ContainerInfoRequest) (*cadvisor
 	return kl.cadvisor.MachineInfo()
 }
+
+func (kl *Kubelet) SyncRocketPods(allPods []api.BoundPod, podSyncTypes map[types.UID]metrics.SyncPodType, start time.Time) error {
```
Why is there rocket specific implementation in the kubelet?
I would expect all this to be in the rocket implementation of `container.Runtime`.

yifan-gu replied:
@sym3tri  It should be.  I should `s/SyncPods/SyncDockerPods` and `s/SyncRocketPods/SyncPod`.

But I am just doing `SyncRocketPods` now to highlight the new code path :D

sym3tri replied:
ok, i think we should do as you suggested before sending the PR up to k8s.

philips replied:
What was suggested before?

sym3tri replied:
@philips I meant what @yifan-gu suggested in his comment about the replacement. https://github.com/coreos/kubernetes/pull/4#discussion_r26350309

Commit:
rocket: support volume mounts.

philips commented on pkg/kubelet/container/runtime_cache.go:
```diff
@@ -0,0 +1,124 @@
+	}, nil
+}
+
+// runtimeCache is an implementation of RuntimeCache interface.
```
the comments on this struct don't add much, I would remove them.

philips commented on pkg/kubelet/container/runtime_cache.go:
```diff
@@ -0,0 +1,124 @@
+	// The content of the cache.
+	pods []*api.Pod
+	// Whether the background thread updating the cache is running.
+	updatingCache bool
```
this can probably just be called updating

philips commented on pkg/kubelet/container/runtime_cache.go:
```diff
@@ -0,0 +1,124 @@
+			return err
+		}
+		r.pods = pods
+		r.cacheTime = time.Now()
```
this should be a helper function as it is identical lines 73 - 78

yifan-gu replied:
@philips Ahh, I just copy-pasted the cache stuff from the master with a different name....
Will do.

philips commented on pkg/kubelet/kubelet.go:
```diff
@@ -1943,3 +1947,222 @@ func (kl *Kubelet) GetRootInfo(req *cadvisorApi.ContainerInfoRequest) (*cadvisor
+	}
+	kl.removeOrphanedStatuses(podFullNames)
+
+	// Filtered out the rejected pod. They don't have running containers.
```
"Create a pod list with the the failed pods filtered out?"

yifan-gu replied:
fixed

philips commented on pkg/kubelet/rocket/rocket.go:
```diff
@@ -0,0 +1,452 @@
+
+const (
+	kubernetesUnitPrefix = "K8S"
+	systemdServiceDir    = "/etc/systemd/system"
```
this should probably go into /run/systemd/system

philips replied:
right @marineam ?

Commit:
rocket: add test for run with mounted volumes.

jonboulle commented on pkg/kubelet/rocket/rocket.go:
```diff
@@ -0,0 +1,452 @@
+const (
+	rocketBinName = "rkt"
+
+	podStateEmbryo    = "embryo"
```
hrm, should these be upstream?

yifan-gu replied:
yes, @jonboulle but it's not exported last time I looked... Should add a TODO here, thx!

Commit:
rocket: change the test aci.

jonboulle commented on cmd/kubelet/app/server.go:
```diff
@@ -129,6 +130,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Var(&s.ClusterDNS, "cluster_dns", "IP address for a cluster DNS server.  If set, kubelet will configure all containers to use this for DNS resolution in addition to the host's DNS servers")
 	fs.BoolVar(&s.ReallyCrashForTesting, "really_crash_for_testing", s.ReallyCrashForTesting, "If true, crash with panics more often.")
 	fs.DurationVar(&s.StreamingConnectionIdleTimeout, "streaming_connection_idle_timeout", 0, "Maximum time a streaming connection can be idle before the connection is automatically closed.  Example: '5m'")
+	fs.StringVar(&s.ContainerRuntimeChoice, "container_runtime_choice", "docker", "The choice fo the container runtime. Example: 'docker', 'rocket'.")
```
"choice" seems superfluous, how about just ContainerRuntime/container_runtime
also s/fo/of/

yifan-gu replied:
@jonboulle Sounds good, but in kubelet there is already a containerRuntime field which stores the interface.
Maybe just change the `container_runtime_choice` to `container_runtime`, and keeps the variable's name?

Hopefully the `containerRuntimeChoice` variable will be removed soon

jonboulle replied:
@yifan-gu sgtm

jonboulle commented on pkg/kubelet/container/runtime_cache.go:
```diff
@@ -0,0 +1,125 @@
+	return nil
+}
+
+// startUpdateingCache continues to invoke ListPods to get the newest result until
```
startUpdatingCache

jonboulle commented on pkg/kubelet/container/runtime_cache.go:
```diff
@@ -0,0 +1,125 @@
+	for run {
+		time.Sleep(defaultUpdateInterval)
+		pods, err := r.runtime.ListPods()
+		cacheTime := time.Now()
```
move after the if err?

jonboulle commented on pkg/kubelet/kubelet.go:
```diff
@@ -1943,3 +1951,222 @@ func (kl *Kubelet) GetRootInfo(req *cadvisorApi.ContainerInfoRequest) (*cadvisor
+
+	glog.V(4).Infof("Desired: %#v", pods)
+	var err error
+	desiredPods := make(map[types.UID]empty)
```
move to 1986

jonboulle commented on pkg/kubelet/kubelet.go:
```diff
@@ -1943,3 +1951,222 @@ func (kl *Kubelet) GetRootInfo(req *cadvisorApi.ContainerInfoRequest) (*cadvisor
+	}
+
+	glog.V(4).Infof("Desired: %#v", pods)
+	var err error
```
redundant

jonboulle commented on pkg/kubelet/kubelet.go:
```diff
@@ -81,6 +84,7 @@ type SyncHandler interface {
 	// Syncs current state to match the specified pods. SyncPodType specified what
 	// type of sync is occuring per pod. StartTime specifies the time at which
 	// syncing began (for use in monitoring).
+	SyncDockerPods(pods []api.BoundPod, podSyncTypes map[types.UID]metrics.SyncPodType, startTime time.Time) error
```
Why don't you just have a docker and non-docker implementation of the original `SyncHandler` interface? Then inject the relevant one to `kl.syncLoop`.

yifan-gu replied:
@jonboulle  Because it's kubelet itself implementing the interface... The interface is just for testability.

jonboulle replied:
Well, that's easy, you just add a kubelet argument to the function. You can leave it like this for now if you want, I just think e.g. the if statement on 1572 is icky and I wouldn't be surprised if they don't like it... but who knows :-)

yifan-gu replied:
Updated that if statement, make it nicer a little bit.

jonboulle commented on pkg/kubelet/container/runtime_cache.go:
```diff
@@ -0,0 +1,125 @@
+func (r *runtimeCache) startUpdatingCache() {
+	run := true
+	for run {
+		time.Sleep(defaultUpdateInterval)
```
is this a kubernetes pattern? Kinda surprised this isn't done with a select/time.After

yifan-gu replied:
@jonboulle Not sure if it's a pattern, just copied with variable names changed... I think we can refactor this later?

jonboulle replied:
@yifan-gu Yes, that's fine; sorry, it's hard to tell what you have written from scratch and what you have copied from elsewhere+reworked.

jonboulle commented on pkg/kubelet/kubelet.go:
```diff
@@ -91,6 +95,7 @@ type volumeMap map[string]volume.Interface
 // New creates a new Kubelet for use in main
 func NewMainKubelet(
 	hostname string,
+	containerRuntimeChoice string,
```
re: line 121/126 (GH won't let me comment down there) - we don't need to wait for Docker if containerRuntimeChoice is rocket, do we?

yifan-gu replied:
@jonboulle Just wrap that with an `if`, I think the logic should be put in a separate func after we merge docker and container runtime together.

jonboulle replied:
ack

jonboulle commented on pkg/kubelet/kubelet.go:
```diff
@@ -1943,3 +1951,222 @@ func (kl *Kubelet) GetRootInfo(req *cadvisorApi.ContainerInfoRequest) (*cadvisor
+	// second one as duplicated and kill it.
+	containersInPod := make(Containers)
+	for i, c := range runningPod.Spec.Containers {
+		containersInPod[i] = &c
```
`c := c` (or `containersInPod[i] = &runningPod.Spec.Containers[i]`)
http://play.golang.org/p/VaDvk0yWP7

yifan-gu replied:
my fault. thx...

jonboulle commented on pkg/kubelet/rocket/rocket.go:
```diff
@@ -0,0 +1,453 @@
+	return result
+}
+
+// New creats the rocket container runtime which implements the container runtime interface.
```
creates

yifan-gu replied:
haha, looks like I am a fan of unix...

jonboulle replied:
:D

jonboulle commented on pkg/kubelet/rocket/rocket.go:
```diff
@@ -0,0 +1,453 @@
+
+// New creats the rocket container runtime which implements the container runtime interface.
+// It will test if the 'rkt' binary is in the $PATH, and whether we can get the
+// version of it. If so, creates the rocket container runtime, other returns an error.
```
s/other/else/

jonboulle commented on pkg/kubelet/rocket/rocket.go:
```diff
@@ -0,0 +1,453 @@
+}
+
+// RunPod first creates the unit file for a pod, and then invokes
+// 'systemctl start' to run that unit file.
```
not strictly true, how about "then calls StartUnit over d-bus"

jonboulle commented on pkg/kubelet/rocket/rocket.go:
```diff
@@ -0,0 +1,453 @@
+	if err != nil {
+		return err
+	}
+	if status := <-ch; status != "done" {
```
a little concerned about blocking here, how do they deal with this in Docker?

yifan-gu replied:
@jonboulle I didn't find particular place they deal with docker, they just wait for the result of `docker.StartContainer`
BTW, what's the difference between `StartUnit` and `StartTransientUnit`? Cannot get much sense after reading the doc...

jonboulle replied:
@yifan-gu Well, `StartUnit` calls must reference a unit that already exists/is known by systemd (i.e. one you've written to disk), whereas with `StartTransientUnit` you can just pass in an entire unit and it's only ever hosted in memory.

Since we don't have any persistence needs (i.e. we're writing the unit files in /run anyway) we _might_ be able to get away with using TransientUnits for everything but there's probably some subtle reason why not that I'm missing right now, so I think just stick with this approach.

yifan-gu replied:
@jonboulle Cool. Thank you.

jonboulle commented on pkg/kubelet/rocket/rocket.go:
```diff
@@ -0,0 +1,453 @@
+
+// RunContainerInPod launches a container in the given pod.
+// For now, we need to kill and restart the whole pod. Hopefully we will be
+// launching this single container without touching its sliblings in the near future.
```
siblings

jonboulle commented on pkg/kubelet/rocket/rocket.go:
```diff
@@ -0,0 +1,453 @@
+// TODO(yifan), when BoundPod and Pod are combined, we can change the input to
+// take just a pod.
+func makePodServiceFileName(podName, podNamespace string) string {
+	return fmt.Sprintf("%s_%s_%s.service", kubernetesUnitPrefix, podName, podNamespace)
```
just to check, we don't have to worry about sanitising the podName/podNamespace?

yifan-gu replied:
@jonboulle Do you mean the string might contain illegal chars like "#$@#" ? Or do you mean there is a concern for duplicated names?

A safer way is to use UID of the pod, but I just made this to be human readable, so friendly for debugging.

I think it's fine now, let me add a TODO.

jonboulle replied:
Yep, just illegal characters, I am not sure what the restrictions on the
podname are

On Fri, Mar 13, 2015 at 4:33 PM, Yifan Gu notifications@github.com wrote:

> In pkg/kubelet/rocket/rocket.go
> https://github.com/coreos/kubernetes/pull/4#discussion_r26430511:
> 
> > -           }
> > -       }
> > -   }
> >   +
> > -   if err = r.getPodStatus(&pod); err != nil {
> > -       glog.Errorf("Error getting pod status of pod %q.", pod.Name)
> > -       return nil, err
> > -   }
> > -   return &pod, nil
> >   +}
> >   +
> >   +// makePodServiceFileName constructs the unit file name for the given pod.
> >   +// TODO(yifan), when BoundPod and Pod are combined, we can change the input to
> >   +// take just a pod.
> >   +func makePodServiceFileName(podName, podNamespace string) string {
> > -   return fmt.Sprintf("%s_%s_%s.service", kubernetesUnitPrefix, podName, podNamespace)
> 
> @jonboulle https://github.com/jonboulle Do you mean the string might
> contain illegal chars like "#$@#" ? Or do you mean there is a concern for
> duplicated names?
> 
> A safer way is to use UID of the pod, but I just made this to be human
> readable, so friendly for debugging.
> 
> I think it's fine now, let me add a TODO.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/coreos/kubernetes/pull/4/files#r26430511.

jonboulle commented on pkg/kubelet/rocket/rocket_test.go:
```diff
@@ -0,0 +1,313 @@
+
+func init() {
+	// Disabled by default since these tests require root privilege.
+	flag.BoolVar(&enableTests, "enable-rocket-tests", false, "Wheter the rocket tests should be enabled")
```
whether

jonboulle commented on pkg/kubelet/runonce.go:
```diff
@@ -61,7 +61,12 @@ func (kl *Kubelet) runOnce(pods []api.BoundPod) (results []RunPodResult, err err
-			err := kl.runPod(pod)
+			var err error
+			if kl.containerRuntimeChoice == "rocket" {
+				err = kl.runRocketPod(pod)
```
any reason why we have 

```
syncPod      <-> syncDockerPod
runRocketPod <-> runPod
```

jonboulle commented on pkg/kubelet/runonce.go:
```diff
@@ -138,3 +143,45 @@ func (kl *Kubelet) isPodRunning(pod api.BoundPod, dockerContainers dockertools.D
+	for {
+		runningPods, err := kl.containerRuntime.ListPods()
+		if err != nil {
+			return fmt.Errorf("failed to get kubelet docker containers: %v", err)
```
needs update

Commit:
kubelet: rename SyncPods and syncPod.

Rename SyncPods to SyncDockerPods.
Rename syncPod to syncDockerPods.
Rename SyncRocketPods to SyncPods.
Rename syncRocketPod to syncPod.

jonboulle commented:
lgtm, go for it!
